README
Net-DHCP version 0.66
=====================

0. DOCS/ AND T/DATA FILES WARNING
----------------------

Some files in the docs/ directory are copied from IANA.
They may not be compatible with the MIT license. I have
included them in the tar ball as-is for the sake of reference.

Currently these files are:
- docs/bootp-dhcp-parameters.txt

Some files in t/data are copied from various sources:
- http://packetlife.net/captures/DHCP_MessageType%2010%2C11%2C12%20and%2013.cap
- http://packetlife.net/captures/DHCP_Inter_VLAN.cap
- http://uluru.ee.unsw.edu.au/~tim/zoo/f994f275.pcap
- https://wiki.wireshark.org/SampleCaptures?action=AttachFile&do=get&target=dhcp-auth.pcap.gz

All other files are licensed as described at the bottom of this file.

Please remove the problem files if needed for distribution packaging etc.

You can fork and submit enhancements etc via https://github.com/djzort/Net-DHCP

1. DESCRIPTION
--------------

Net::DHCP is a DHCP set of classes designed to handle basic DHCP
handling. It can be used to develop either client, server or relays.
It is composed of 100% pure Perl.

The author invites feedback on Net::DHCP. If there's something you'd
like to have added, please let me know.  If you find a bug, please
send me the information described in the BUGS section below.

The original version of this module was written by Francis van Dun,
and has been deeply reorganized.

2. PREREQUISITES
----------------

Net::DHCP requires Perl 5.8.0 or higher.

3. INSTALLATION
---------------

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

4. BUGS
-------

If you find a bug, please report it to the author along with the
following information:

    * version of Perl (output of 'perl -V' is best)
    * version of Net::DHCP
    * operating system type and version
    * exact text of error message or description of problem
    * the shortest possible program that exhibits the problem

COPYRIGHT AND LICENCE

Copyright (C) 2002 by Francis van Dun.
Copyright (C) 2005,2006 by Stephan Hadinger.
Copyright (C) 2010-2022 by Dean Hamstead.

The MIT License

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software
without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to
whom the Software is furnished to do so, subject to the
following conditions:

The above copyright notice and this permission notice shall
be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT
WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE AND NONINFRINGEMENT. IN NO EVENT
SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.


Changes
Net::DHCP - A DHCP packet object

0.7      2022-06-22
 	- Increase minimum perl to 5.12 to ensure Socket is new enough for Net::Frame
	- Update Changes to provide more details
	- Bump to 0.7
	- WARNING! This release might break your toys. Please test it first before upgrading!

0.7_010  2022-06-22
	- Update license to MIT, approved by all contributors https://github.com/djzort/Net-DHCP/issues/11

0.7_009  2022-06-22
	- Accepted PR https://github.com/djzort/Net-DHCP/pull/10 - Implement levels of packet minimal length checking

0.7_008  2018-02-16
	- Merge PR https://github.com/djzort/Net-DHCP/pull/8 -  Declare provided modules in meta.{yml,json}

0.7_007  2016-07-25
	- Merge PR https://github.com/djzort/Net-DHCP/pull/6 - Can use addSubOptionValue for some DHO_DHCP_AGENT_OPTIONS

0.7_006  2016-02-24
	- Minor change via https://github.com/djzort/Net-DHCP/pull/5

0.7_005  2015-09-21
	- Fix build deps for BZ107212

0.7_004  2015-09-15
	- Some basic support for option 120

0.7_003  2015-08-28
	- Add more test examining reeal packets
	- Start handling client id's
	- Add author covenant
	- Moved various methods to Net::DHCP::Packet::Attributes
	- Moved various methods to Net::DHCP::Packet::IPv4Utils
	- Adjusted tests accordingly
	- Net::DHCP no longer exports unpackinet, packinet, unpackinets, packinets

0.7_002  2015-08-15
	- Add tests examining real packets from tcpdump

0.7_001  2015-08-08
	- Reworked ->new()
	- Bring back Net/DHCP.pm
	- Add Virtual Subnet Selection constant from rfc 6607
	- Add tests examining real packets from tcpdump
	- Option 82 decoding
	- Lists of values are now retured with commas

0.696  2015-06-10
	- Fixed up dist.ini
	- YAPC::NA release

0.695  2014-12-02
	- Corrected github details

0.694  2014-08-27

	- Fix for RT89267
	- Added github details
	- Some minor code style changes

0.693  2012-10-16

	- Changes kindly sent in my Alexander Clouter
	- Allows static routes to be included in dhcp packets
	- (no tests yet)

0.692  2012-09-05

	- Minor typo in pod

0.691  2012-08-24

	- Updated changelog for Kawlitee

0.69  2012-06-06

	- Added 'use warnings' to improve Kwalitee
	- Now uses dist::zilla

0.685  2012-05-21

	- Removed file

0.684  2012-05-21

	- Changed to using Dist::Zilla.

0.683  2012-01-21

	- Still chasing the Build requires issue.

0.682  2012-01-21

	- Minor update to tests sent in from Gregory Herrmann via RT#74187. Thanks for that!
	- Updated pod to complete coverage

0.681  2012-01-16

	- Forgot to update MANIFEST file... woops. Also some pod updates.

0.68   2012-01-14

	- Added Net/DHCP.pm stub.

0.67_2  2011-04-18

	- Updated pod. Please be awair that i have added the word "CableLabs" to to pod. This 05-be of concern to Debian. Im not sure how their licensing works exactly vs CableLabs copyrights. Im reasonably certain its aok.
	- Minor changes for RT#73972
	- Improvements to tests

0.67_1  2010-12-16

	- Warning: the declaration of DHO_NIS_DOMAIN was overlapping for options 40 options 64. Therefore, options 64 and 65 where renamed DHO_NISV3_SERVER and DHO_NISV3_DOMAIN to remove this overlap and remain consistent.
	- Constants.pm. Added a vast number of option codes from the IANA listing.
	- Added updated DHCP message codes, added Netware IP Codes. Tests were added for coverage of constants.
	- Packet.pm. Changed a whole bunch of stuff. Including removing some then then else orgies, tidying up lots of warnings.
	- Tests. I'd added a bunch more tests :)
	- Rearranged the test suite. Added lots of tests.

0.67  2010-11-13

	- Fixing a small bug in marshall (see cpanrt61693), fixed a bug in Constants.pm (see cpanrt32962), clarified version dependance closing out cpanrt19405

0.66  2006-07-12

	- Module is definitely not compatible with Perl 5.6.0. Minimum version is Perl 5.8.0.
	- Corrected major bug in removeOption(), thanks to Nathan Scarlett.

0.65  2006--04-22

	- Added a sample DHCP Server and tester: dhcpd.pl & dhcpd_test.pl in the examples folder.

0.64  2006-03-06

	- Added removeOption() function

0.63  2006-01-01

	- Major bug corrected in Net::DHCP::Packet->new() while parsing attributes.

0.62  2005-10-07

	- Minor change to make module compatible Perl 5.6.0 instead of only 5.8.0.

0.61  2005-04-29

	- Changed minimal acceptable packet length in marshall() method.

0.60  2005-03-29

	- Lots of tests added.
	- getOption() and addOption() removed and replaced by:
	  * getOptionValue() & addOptionValue() which provide automatic
	  type conversion to/from text representation
	  * getOptionRaw() & addOptionRaw() to manipulate binary format packed into strings.
	- new() now behaves differently for DHCP options, requiring text values and not packed binary values.
	- toString() now escapes unprintable characters.
	- Added methods : ciaddrRaw, giaddrRaw, siaddrRaw, yiaddrRaw that use packed binary formats. They are provided for performance enhancements when you want to avoid unnecessary type conversion.
	- Pad options (0x00) are now supported.
	- Example of DHCPLeaseQuery provided by John A. Murphy.

0.53  2005-03-02

	- Performance enhancement in Packet->new() avoirding unnecessary type conversion and function calls.
	- Support for DHCPLEASEQUERY, Cisco extension (draft-ietf-dhc-leasequery-05.txt).
	- Corrected bug in Packet->new() not allowing '0' in Hlen and Htype parameters.  Thanks to John A. Murphy.

0.52  2005-04-24

	- Performance enhancements in Packet.pm.

0.51  2005-04-20

	- Added encoding and decoding functions for Relay Agent information (option 82 - RFC 3046).
	- Comment field added in the Packet object. This lets you put any comment in the Packet object, such as a transaction number. This field not processed.

0.50  2005-04-8

	- Deeply rewritten version based on Net::DHCP v0.11 by Francis van Dun.
	- Net::DHCP::Session is momentarily missing, it should be reintroduced soon.


lib/Net/DHCP/Packet/OrderOptions.pm
#!/bin/false
# PODNAME: Net::DHCP::Packet::OrderOptions
# ABSTRACT: Option ordering logic for Net::DHCP
use strict;
use warnings;
use 5.8.0;

package Net::DHCP::Packet::OrderOptions;

# standard module declaration
our ( @ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS );
use Exporter;
@ISA       = qw(Exporter);
@EXPORT    = qw( );
@EXPORT_OK = qw(
);
%EXPORT_TAGS = ( all => \@EXPORT_OK );

use Carp qw/ carp /;

my @order = (

# taken from isc-dhcp... i think its wrong though
53, # message type
54, # server id
51, # lease time
58, # renewal time
59, # rebinding time
56, # message
50, # requested address
92, # associated ip
118, # subnet selection
61, # client id
81, # fqdn
1, # subnet mask
3, # routers
6, # domain name servers
12, # host name

2,
4,
5,
7,
8,
9,
10,
11,
13,
14,
15,
16,
17,
18,
19,
20,
21,
22,
23,
24,
25,
26,
27,
28,
29,
30,
31,
32,
33,
34,
35,
36,
37,
38,
39,
40,
41,
42,
44,
45,
46,
47,
48,
49,
52,
55,
57,
60,
62,
63,
64,
65,
66,
67,
68,
69,
70,
71,
72,
73,
74,
75,
76,
77,
78,
79,
80,
83,
84,
85,
86,
87,
88,
89,
90,
91,
93,
94,
95,
96,
97,
98,
99,
100,
101,
102,
103,
104,
105,
106,
107,
108,
109,
110,
111,
112,
113,
114,
115,
116,
117,
119,
120,
121,
122,
123,
124,
125,
126,
127,
128,
129,
130,
131,
132,
133,
134,
135,
136,
137,
138,
139,
140,
141,
142,
143,
144,
145,
146,
147,
148,
149,
150,
151,
152,
153,
154,
155,
156,
157,
158,
159,
160,
161,
162,
163,
164,
165,
166,
167,
168,
169,
170,
171,
172,
173,
174,
175,
176,
177,
178,
179,
180,
181,
182,
183,
184,
185,
186,
187,
188,
189,
190,
191,
192,
193,
194,
195,
196,
197,
198,
199,
200,
201,
202,
203,
204,
205,
206,
207,
208,
209,
210,
211,
212,
213,
214,
215,
216,
217,
218,
219,
220,
221,
222,
223,
224,
225,
226,
227,
228,
229,
230,
231,
232,
233,
234,
235,
236,
237,
238,
239,
240,
241,
242,
243,
244,
245,
246,
247,
248,
249,
250,
251,
252,
253,
254,

# intel pxe really wants 60 before 43
43,

# Cablelabs really really wants 82 at the end
82,

);

#=======================================================================
# comment attribute : enables transaction number identification
#
#

sub optionsorder {



}

#=======================================================================

1;

=pod

=head1 SYNOPSIS

   use Net::DHCP::Packet::OrderOptions qw( :all );

=head1 DESCRIPTION

Provides sorting for optons in Net::DHCP::Packet.

This module is not particularly useful on its own.

Because DHCP clients generally suck, the order of options in the dhcp
packet actually matter. This module keeps the handling of all that
quirky madness in one place

=head1 QUIRK WORK-AROUNDS

Cable vendors really want option 82 to always be last.

Intel PXE really wants option 60 before option 43.

Minimum frame size is 300, but we only warn (carp) if its not.

=head1 METHODS

=over 4

=item optionsorder ( )

TODO This will rearrange the order of options to accomidate as many
quirky cliens as possible.

=back

=head1 SEE ALSO

L<Net::DHCP>, L<Net::DHCP::Packet>, L<Net::DHCP::Constants>,
L<dhcp-options>.

=cut
